In [1]:
###############################################################################
# The Institute for the Design of Advanced Energy Systems Integrated Platform
# Framework (IDAES IP) was produced under the DOE Institute for the
# Design of Advanced Energy Systems (IDAES).
#
# Copyright (c) 2018-2023 by the software owners: The Regents of the
# University of California, through Lawrence Berkeley National Laboratory,
# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
# University, West Virginia University Research Corporation, et al.
# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
# for full copyright and license information.
###############################################################################

Diagnostics Tutorial: Debugging a Structural Singularity
===========================
Author: Robert Parker\
Maintainer: Robert Parker\
Updated: 2024-05-24

In this tutorial, we will use the [IDAES Diagnostics Toolbox](https://idaes-pse.readthedocs.io/en/2.4.0/explanations/model_diagnostics/index.html#diagnostics-toolbox)
to diagnose and fix a structural singularity that is preventing us from solving an optimization problem.

## Construct our model

Suppose a collaborator has given us a model to work with. They give us a square model and tell us what the degrees of freedom are. We construct an optimization problem and try to solve it. In this tutorial, we don't want to worry too much about the details that go into constructing the model. This has been provided in the `idaes_examples.mod.diagnostics.gas_solid_contactors.model` module.

### Model details (OKAY TO SKIP)

The model we are trying to optimize is a dynamic model of a moving bed chemical looping combustion reactor. The model has been described by [Okoli et al.][1] and [Parker and Biegler][2]. This is a gas-solid reactor with counter-current flow. The degrees of freedom are gas and solid inlet flow rates, and we are trying to minimize the deviation from a desired operating point via a least-squares objective function.

[1]: https://www.sciencedirect.com/science/article/pii/S0032591019302803
[2]: https://www.sciencedirect.com/science/article/pii/S2405896322008825

Again, we don't want to worry too much about the model. The `make_model` function will construct the optimization problem that we want to solve, and whenever we do something model-specific, we will explicitly make note of it.

# Trying to solve the original model

With that out of the way, let's construct the model and try to solve it!

In [2]:
from idaes_examples.mod.diagnostics.gas_solid_contactors.model import make_model

# This constructs a dynamic model with degrees of freedom and an objective function.
model = make_model()

2024-05-24 17:39:12 [WARNING] idaes.core.base.property_meta: DEPRECATED: The property name diffusion_comp in property metadata is
not a recognized standard property name defined in this PropertySet.
Please refer to IDAES standard names in the IDAES documentation. You
can use the define_custom_properties() rather than the
add_properties() method to define metadata for this property. You can
also use a different property set by calling the define_property_set()
method.  (deprecated in 2.0.0, will be removed in (or after) 3.0.0)
(called from /Users/rbparker/research/upstream-dev/idaes-examples/idaes_examples/mod/diagnostics/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:202)


'pyomo.core.base.var.ScalarVar'>) on block fs.MB with a new Component
(type=<class 'pyomo.core.base.var.AbstractScalarVar'>). This is usually
indicative of a modelling error. To avoid this warning, use
block.del_component() and block.add_component().


(type=<class 'pyomo.core.base.constraint.ScalarConstraint'>) on block fs.MB
with a new Component (type=<class
'pyomo.core.base.constraint.AbstractScalarConstraint'>). This is usually
indicative of a modelling error. To avoid this warning, use
block.del_component() and block.add_component().


(type=<class 'pyomo.core.base.constraint.ScalarConstraint'>) on block fs.MB
with a new Component (type=<class
'pyomo.core.base.constraint.AbstractScalarConstraint'>). This is usually
indicative of a modelling error. To avoid this warning, use
block.del_component() and block.add_component().


Constructing a steady model to initialize the dynamic model
'pyomo.core.base.var.ScalarVar'>) on block fs.MB with a new Component
(type=<class 'pyomo.core.base.var.AbstractScalarVar'>). This is usually
indicative of a modelling error. To avoid this warning, use
block.del_component() and block.add_component().


(type=<class 'pyomo.core.base.constraint.ScalarConstraint'>) on block fs.MB
with a new Component (type=<class
'pyomo.core.base.constraint.AbstractScalarConstraint'>). This is usually
indicative of a modelling error. To avoid this warning, use
block.del_component() and block.add_component().


(type=<class 'pyomo.core.base.constraint.ScalarConstraint'>) on block fs.MB
with a new Component (type=<class
'pyomo.core.base.constraint.AbstractScalarConstraint'>). This is usually
indicative of a modelling error. To avoid this warning, use
block.del_component() and block.add_component().


Initializing steady model
2024-05-24 17:39:12 [INFO] idaes.init.fs.MB: Initialize Thermophysical Properties


2024-05-24 17:39:12 [INFO] idaes.init.fs.MB: Initialize Hydrodynamics


2024-05-24 17:39:12 [INFO] idaes.init.fs.MB: Initialize Mass Balances


2024-05-24 17:39:13 [INFO] idaes.init.fs.MB: Initialize Energy Balances


Ipopt 3.13.2: 

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Process Systems Engineering Framework (IDAES PSE
    Framework) Copyright (c) 2018-2019. See https://github.com/IDAES/idaes-pse.

This version of Ipopt was compiled using HSL, a collection of Fortran codes
    for large-scale scientific computation.  All technical papers, sales and
    publicity material resulting from use of the HSL codes within IPOPT must
    contain the following acknowledgement:
        HSL, a collection of Fortran codes for large-scale scientific
        computation. See http://

Solving square problem with dynamic model


Ipopt 3.13.2: 

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Process Systems Engineering Framework (IDAES PSE
    Framework) Copyright (c) 2018-2019. See https://github.com/IDAES/idaes-pse.

This version of Ipopt was compiled using HSL, a collection of Fortran codes
    for large-scale scientific computation.  All technical papers, sales and
    publicity material resulting from use of the HSL codes within IPOPT must
    contain the following acknowledgement:
        HSL, a collection of Fortran codes for large-scale scientific
        computation. See http://

Cannot recompute multipliers for feasibility problem.  Error in eq_mult_calculator

Number of Iterations....: 0

                                   (scaled)                 (unscaled)
Objective...............:   0.0000000000000000e+00    0.0000000000000000e+00
Dual infeasibility......:   0.0000000000000000e+00    0.0000000000000000e+00
Constraint violation....:   9.3496055342257023e-10    1.4901161193847656e-08
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   9.3496055342257023e-10    1.4901161193847656e-08


Number of objective function evaluations             = 1
Number of objective gradient evaluations             = 1
Number of equality constraint evaluations            = 1
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 1
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 0
Total CPU secs in IPOPT (w/o f

'pyomo.core.base.var.ScalarVar'>) on block fs.MB with a new Component
(type=<class 'pyomo.core.base.var.AbstractScalarVar'>). This is usually
indicative of a modelling error. To avoid this warning, use
block.del_component() and block.add_component().


(type=<class 'pyomo.core.base.constraint.ScalarConstraint'>) on block fs.MB
with a new Component (type=<class
'pyomo.core.base.constraint.AbstractScalarConstraint'>). This is usually
indicative of a modelling error. To avoid this warning, use
block.del_component() and block.add_component().


(type=<class 'pyomo.core.base.constraint.ScalarConstraint'>) on block fs.MB
with a new Component (type=<class
'pyomo.core.base.constraint.AbstractScalarConstraint'>). This is usually
indicative of a modelling error. To avoid this warning, use
block.del_component() and block.add_component().


Initializing steady model
2024-05-24 17:39:14 [INFO] idaes.init.fs.MB: Initialize Thermophysical Properties


2024-05-24 17:39:14 [INFO] idaes.init.fs.MB: Initialize Hydrodynamics


2024-05-24 17:39:14 [INFO] idaes.init.fs.MB: Initialize Mass Balances


2024-05-24 17:39:15 [INFO] idaes.init.fs.MB: Initialize Energy Balances


Ipopt 3.13.2: 

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Process Systems Engineering Framework (IDAES PSE
    Framework) Copyright (c) 2018-2019. See https://github.com/IDAES/idaes-pse.

This version of Ipopt was compiled using HSL, a collection of Fortran codes
    for large-scale scientific computation.  All technical papers, sales and
    publicity material resulting from use of the HSL codes within IPOPT must
    contain the following acknowledgement:
        HSL, a collection of Fortran codes for large-scale scientific
        computation. See http://

Ipopt 3.13.2: 

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Process Systems Engineering Framework (IDAES PSE
    Framework) Copyright (c) 2018-2019. See https://github.com/IDAES/idaes-pse.

This version of Ipopt was compiled using HSL, a collection of Fortran codes
    for large-scale scientific computation.  All technical papers, sales and
    publicity material resulting from use of the HSL codes within IPOPT must
    contain the following acknowledgement:
        HSL, a collection of Fortran codes for large-scale scientific
        computation. See http://

Before trying to solve the model, let's make sure it conforms to our expectiations, i.e. it (a) has degrees of freedom and (b) is well-initialized to a feasible point.

In [3]:
# Import some useful utilities from the model_statistics module.
# Degrees of freedom and constraint residuals are always good things to check before
# trying to solve a simulation or optimization problem.
from idaes.core.util.model_statistics import degrees_of_freedom, large_residuals_set

dof = degrees_of_freedom(model)
print(f"Degrees of freeom: {dof}")
has_large_residuals = bool(large_residuals_set(model, tol=1e-5))
print(f"Has large residuals: {has_large_residuals}")

Degrees of freeom: 10
Has large residuals: False


Looks good so far. Let's try to solve.

In [4]:
# Import pyomo.environ for access to solvers
import pyomo.environ as pyo

solver = pyo.SolverFactory("ipopt")
solver.options["max_iter"] = 20
solver.options["print_user_options"] = "yes"
solver.options["OF_print_info_string"] = "yes"
res = solver.solve(model, tee=True)
print(f"Converged successfully: {pyo.check_optimal_termination(res)}")

Ipopt 3.13.2: max_iter=20
print_user_options=yes
option_file_name=/var/folders/j3/k8s7mg5x053c0rgp8sd87d540016lk/T/tmpzo3fzubx_ipopt.opt



Using option file "/var/folders/j3/k8s7mg5x053c0rgp8sd87d540016lk/T/tmpzo3fzubx_ipopt.opt".


List of user-set options:

                                    Name   Value                used
                                max_iter = 20                    yes
                        option_file_name = /var/folders/j3/k8s7mg5x053c0rgp8sd87d540016lk/T/tmpzo3fzubx_ipopt.opt  yes
                       print_info_string = yes                   yes
                      print_user_options = yes                   yes

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Pro

   1  1.7330638e+04 8.08e+01 3.86e+03  -1.0 5.38e+02  -4.0 1.00e+00 1.25e-01f  4 Nh LTmaxTmaxTmaxTmax


   2  1.7264117e+04 2.62e+02 2.39e+04  -1.0 2.13e+01  -1.8 1.00e+00 1.00e+00f  1 LMcqaDj L


   3  1.7284985e+04 9.61e+01 1.96e+06  -1.0 5.31e+00   0.5 1.00e+00 1.00e+00h  1 l


   4  1.7289252e+04 1.01e+02 1.07e+06  -1.0 2.23e+00   1.8 1.00e+00 1.00e+00h  1 l


   5  1.7280273e+04 4.18e+02 4.82e+06  -1.0 1.42e+01   1.3 1.00e+00 1.00e+00h  1 l


   6  1.7290534e+04 1.04e+02 7.41e+06  -1.0 1.74e+00   2.6 1.00e+00 1.00e+00h  1 l


   7  1.7292613e+04 1.04e+02 1.56e+06  -1.0 1.46e+00   3.1 1.00e+00 1.00e+00h  1 l


   8  1.7292564e+04 1.04e+02 4.19e+05  -1.0 6.72e-01   3.5 1.00e+00 1.00e+00h  1 l


   9  1.7293810e+04 1.04e+02 8.35e+05  -1.0 3.56e-01   3.0 1.00e+00 1.00e+00h  1 l


iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  10  1.7294764e+04 1.07e+02 4.21e+05  -1.0 1.10e+00   2.5 1.00e+00 1.00e+00h  1 l


  11  1.7294347e+04 1.07e+02 4.09e+05  -1.0 2.10e+01   2.1 1.00e+00 3.12e-02f  6 lTmax


  12  1.7295994e+04 1.10e+02 2.48e+05  -1.0 2.51e+00   2.5 1.00e+00 1.00e+00h  1 l


  13  1.7300256e+04 1.14e+02 1.02e+06  -1.0 3.48e+00   2.0 1.00e+00 1.00e+00h  1 l


  14  1.7301095e+04 1.13e+02 2.27e+05  -1.0 1.39e+00   2.4 1.00e+00 1.00e+00h  1 l


  15  1.7301247e+04 1.12e+02 1.48e+05  -1.0 5.80e-01   2.9 1.00e+00 1.00e+00h  1 l


  16  1.7301288e+04 1.11e+02 7.81e+04  -1.0 2.39e-01   3.3 1.00e+00 1.00e+00h  1 l


  17  1.7301292e+04 1.11e+02 7.71e+04  -1.0 1.55e+00   2.8 1.00e+00 1.56e-02h  7 l


  18  1.7301241e+04 1.10e+02 1.70e+05  -1.0 2.93e-01   3.2 1.00e+00 1.00e+00h  1 l


  19  1.7300685e+04 1.07e+02 9.60e+05  -1.0 9.24e-01   2.8 1.00e+00 1.00e+00h  1 l


iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  20  1.7299590e+04 9.97e+01 5.31e+06  -1.0 2.72e+00   2.3 1.00e+00 1.00e+00f  1 l

Number of Iterations....: 20

                                   (scaled)                 (unscaled)
Objective...............:   1.7299590151685607e+04    1.7299590151685607e+04
Dual infeasibility......:   5.3129041630973332e+06    5.3129041630973332e+06
Constraint violation....:   1.4006242768427571e-01    9.9670794793864445e+01
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   4.5783849858501395e+04    5.3129041630973332e+06


Number of objective function evaluations             = 34
Number of objective gradient evaluations             = 21
Number of equality constraint evaluations            = 39
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 21
Number of inequality constraint Jacobian evaluations = 0
N

model.name="unknown";
    - termination condition: maxIterations
    - message from solver: Ipopt 3.13.2\x3a Maximum Number of Iterations
      Exceeded.


Converged successfully: False


IPOPT fails to solve the optimization problem... You can try increasing the iteration limit, but it is very unlikely that this model will ever solve. A telltale sign that something is wrong with our model is the persistence of regularization coefficients, that is, numbers in the `lg(rg)` column of the IPOPT log. These coefficients can have multiple causes. One is that the constraint Jacobian (partial derivative matrix) is singular, which indicates a problem with our model. We have set the `print_info_string` option in Ipopt to display "diagnostic tags" to help interpret these regularization coefficients. The "L" and "l" diagnostic tags, which appear repeatedly, indicate that the Jacobian is singular. For more information on IPOPT diagnostic tags, see the IPOPT [documentation](https://coin-or.github.io/Ipopt/OUTPUT.html).

# Debugging the original model

Let's run the diagnostics toolbox on the model and see what it has to say.

For good practice, we'll first make sure the model we're debugging is square. Remember that we're assuming we already know how to toggle degrees of freedom in our model.

In [5]:
# Fix gas and solid flow rates at their respective inlets
model.fs.MB.gas_phase.properties[:, 0].flow_mol.fix()
model.fs.MB.solid_phase.properties[:, 1].flow_mass.fix()
# Part of our optimization problem was a set of constraints to enforce piecewise
# constant control inputs. We need to deactivate these as well.
model.piecewise_constant_constraints.deactivate()

Now we can run the diagnostics toolbox.

In [6]:
from idaes.core.util.model_diagnostics import DiagnosticsToolbox

# Before running the diagnostics toolbox, we'll effectively disable Pyomo logging messages.
# This is not recommended in general, but we do it here to suppress unit inconsistency errors
# that otherwise flood our console (and significantly slow down this notebook). This model
# has unit inconsistencies as it was created in IDAES 1.7, before we enforced that models
# use units.
import logging
logging.getLogger("pyomo").setLevel(logging.CRITICAL)

dt = DiagnosticsToolbox(model)
dt.report_structural_issues()

Model Statistics

        Activated Blocks: 387 (Deactivated: 0)
        Free Variables in Activated Constraints: 10200 (External: 0)
            Free Variables with only lower bounds: 0
            Free Variables with only upper bounds: 0
            Free Variables with upper and lower bounds: 0
        Fixed Variables in Activated Constraints: 212 (External: 0)
        Activated Equality Constraints: 10200 (Deactivated: 10)
        Activated Inequality Constraints: 0 (Deactivated: 0)
        Activated Objectives: 1 (Deactivated: 0)

------------------------------------------------------------------------------------
3 WARNINGS

        Under-Constrained Set: 130 variables, 120 constraints
        Over-Constrained Set: 61 variables, 71 constraints

------------------------------------------------------------------------------------
1 Cautions

    Caution: 143 unused variables (3 fixed)

------------------------------------------------------------------------------------
Suggested nex

Let's look at the warnings we got:
- Inconsistent units
- Structural singularity
- Potential evaluation errors

We'll ignore the inconsistent units. The property package and unit model here were extracted from IDAES 1.7, before we rigorously enforced that all models use units. Potential evaluation errors may be worth looking into, but looking at the solver log above, we don't notice many evaluation errors. The structural singularity looks like the most promising avenue to debug, especially as the Ipopt log displays persistent regularization coefficients that appear to be caused by a singular Jacobian.

Let's follow the toolbox's advice and display the under and over-constrained sets.

In [7]:
dt.display_underconstrained_set()
dt.display_overconstrained_set()

Dulmage-Mendelsohn Under-Constrained Set

    Independent Block 0:

        Variables:

            fs.MB.solid_phase.properties[0.0,0.0].flow_mass
            fs.MB.solid_phase.properties[0.0,0.1].flow_mass
            fs.MB.solid_phase.properties[0.0,0.2].flow_mass
            fs.MB.solid_phase.properties[0.0,0.3].flow_mass
            fs.MB.solid_phase.properties[0.0,0.4].flow_mass
            fs.MB.solid_phase.properties[0.0,0.5].flow_mass
            fs.MB.solid_phase.properties[0.0,0.6].flow_mass
            fs.MB.solid_phase.properties[0.0,0.7].flow_mass
            fs.MB.solid_phase.properties[0.0,0.8].flow_mass
            fs.MB.solid_phase.properties[0.0,0.9].flow_mass
            fs.MB.solid_phase._flow_terms[0.0,0.0,Sol,Fe2O3]
            fs.MB.solid_phase._flow_terms[0.0,0.0,Sol,Fe3O4]
            fs.MB.solid_phase._flow_terms[0.0,0.0,Sol,Al2O3]
            fs.MB.solid_phase._enthalpy_flow[0.0,0.0,Sol]
            fs.MB.solid_phase.material_flow_dx[0.0,0.0,Sol,Fe2O3]
     

Dulmage-Mendelsohn Over-Constrained Set

    Independent Block 0:

        Variables:

            fs.MB.solid_phase.properties[0.0,0.0].mass_frac_comp[Fe3O4]
            fs.MB.solid_phase.properties[0.0,0.0].mass_frac_comp[Al2O3]
            fs.MB.solid_phase.properties[0.0,0.0].mass_frac_comp[Fe2O3]
            fs.MB.solid_phase.properties[0.0,0.0].dens_mass_skeletal
            fs.MB.solid_phase.area[0.0,0.0]
            fs.MB.solid_phase.properties[0.0,0.0].dens_mass_particle
            fs.MB.bed_area
            fs.MB.solid_phase.properties[0.0,0.1].mass_frac_comp[Fe3O4]
            fs.MB.solid_phase.properties[0.0,0.1].mass_frac_comp[Al2O3]
            fs.MB.solid_phase.properties[0.0,0.1].mass_frac_comp[Fe2O3]
            fs.MB.solid_phase.properties[0.0,0.1].dens_mass_skeletal
            fs.MB.solid_phase.area[0.0,0.1]
            fs.MB.solid_phase.properties[0.0,0.1].dens_mass_particle
            fs.MB.solid_phase.properties[0.0,0.2].mass_frac_comp[Fe3O4]
            fs.MB.

## Over and under-constrained subsystems

Structural singularities are characterized by the [Dulmage-Mendelson decomposition][3], which partitions a system into minimal over and under-constrained subsystems. These subsystems contain the potentially unmatched constraints and variables, respectively. Here, "unmatched" effectively means "causing a singularity". [Pothen and Fan][4] give a good overview of the Dulmage-Mendelsohn decomposition and [Parker et al.][5] give several examples.

[3]: https://www.cambridge.org/core/journals/canadian-journal-of-mathematics/article/coverings-of-bipartite-graphs/413735C5888AB542B92D0C4F402800B1
[4]: https://dl.acm.org/doi/10.1145/98267.98287
[5]: https://www.sciencedirect.com/science/article/pii/S0098135423002533

The most straightforward way to fix a structural singularity is to fix variables that are in the under-constrained system and deactivate constraints in the over-constrained subsystem. However, this may not be applicable for every model. For example, we may need to add variables and constraints instead. What over and under-constrained subsystems are telling us is that something is wrong with our modeling assumptions. The particular fix that is appropriate will depend heavily on the model.

If the above output gives us any clues, we can go ahead and start trying to fix things. However, suppose it doesn't. A good strategy is to try to break down the model into smaller, square subsystems that we think should be nonsingular. For a dynamic model like this one, a good candidate is the subsystem of variables and equations at each point in time.

In [8]:
# We've included a utility function to extract the subsystem of variables and equations
# at a specified point in time. If you are dealing with a process flowsheet, here you
# may want to exact each unit model individually.
from idaes_examples.mod.diagnostics.util import get_subsystem_at_time
from pyomo.util.subsystems import TemporarySubsystemManager

# Let's start with t=0. Really, we'd probably want to do this in a loop and try all time points.
t0 = model.fs.time.first()
t_block, inputs = get_subsystem_at_time(model, model.fs.time, t0)
# We'll temporarily fix the "inputs" to make sure we have a square system while debugging
with TemporarySubsystemManager(to_fix=inputs):
    dt = DiagnosticsToolbox(t_block)
    dt.report_structural_issues()
    dt.display_underconstrained_set()
    dt.display_overconstrained_set()

Model Statistics

        Activated Blocks: 1 (Deactivated: 0)
        Free Variables in Activated Constraints: 847 (External: 847)
            Free Variables with only lower bounds: 0
            Free Variables with only upper bounds: 0
            Free Variables with upper and lower bounds: 0
        Fixed Variables in Activated Constraints: 103 (External: 103)
        Activated Equality Constraints: 847 (Deactivated: 0)
        Activated Inequality Constraints: 0 (Deactivated: 0)
        Activated Objectives: 0 (Deactivated: 0)

------------------------------------------------------------------------------------
3 WARNINGS

        Under-Constrained Set: 130 variables, 120 constraints
        Over-Constrained Set: 60 variables, 70 constraints

------------------------------------------------------------------------------------
0 Cautions

    No cautions found!

------------------------------------------------------------------------------------
Suggested next steps:

    display_co

Dulmage-Mendelsohn Over-Constrained Set

    Independent Block 0:

        Variables:

            fs.MB.solid_phase.properties[0.0,0.0].mass_frac_comp[Fe3O4]
            fs.MB.solid_phase.properties[0.0,0.0].mass_frac_comp[Al2O3]
            fs.MB.solid_phase.properties[0.0,0.0].mass_frac_comp[Fe2O3]
            fs.MB.solid_phase.properties[0.0,0.0].dens_mass_skeletal
            fs.MB.solid_phase.area[0.0,0.0]
            fs.MB.solid_phase.properties[0.0,0.0].dens_mass_particle

        Constraints:

            fs.MB.solid_phase.material_holdup_calculation[0.0,0.0,Sol,Fe3O4]
            fs.MB.solid_phase.material_holdup_calculation[0.0,0.0,Sol,Al2O3]
            fs.MB.solid_phase.properties[0.0,0.0].sum_component_eqn
            fs.MB.solid_phase.properties[0.0,0.0].density_particle_constraint
            fs.MB.solid_phase_area[0.0,0.0]
            fs.MB.solid_phase.material_holdup_calculation[0.0,0.0,Sol,Fe2O3]
            fs.MB.solid_phase.properties[0.0,0.0].density_skeletal_cons

These over and under-constrained subsystems aren't much smaller, but now the over-constrained system decomposes into 10 small, independent blocks. These should be easier to debug.

## Debugging the over-constrained subsystem

To debug the over-constrained subsystem, we look for a constraint that is not calculating any of the variables in the subsystem. The "odd constraint out" here seems to be the mass fraction sum, `sum_component_eqn`. This must "solve for" one of the mass fractions, which means one of the `material_holdup_calculation` equations must "solve for" particle density rather than mass fraction. If we want to see what variables are contained in one of these constraints, we can always `pprint` it:

In [9]:
model.fs.MB.solid_phase.material_holdup_calculation[0, 0.9, "Sol", "Fe3O4"].pprint()

{Member of material_holdup_calculation} : Material holdup calculations
    Size=363, Index=fs._time*fs.MB.solid_length_domain*fs.solid_properties._phase_component_set, Active=True
    Key                        : Lower : Body                                                                                                                                                                                                            : Upper : Active
    (0.0, 0.9, 'Sol', 'Fe3O4') :   0.0 : fs.MB.solid_phase.material_holdup[0.0,0.9,Sol,Fe3O4] - fs.MB.solid_phase.area[0.0,0.9]*1*(fs.MB.solid_phase.properties[0.0,0.9].dens_mass_particle*fs.MB.solid_phase.properties[0.0,0.9].mass_frac_comp[Fe3O4]) :   0.0 :   True


If one of these `material_holdup_calculation` equations is solving for particle density, then that means that `density_particle_constraint` is not actually solving for density.  Maybe `density_particle_constraint` is over-determining our system?

In [10]:
model.fs.MB.solid_phase.properties[0, 0.9].density_particle_constraint.pprint()

density_particle_constraint : Size=1, Index=None, Active=True
    Key  : Lower : Body                                                                                                                                                            : Upper : Active
    None :   0.0 : fs.MB.solid_phase.properties[0.0,0.9].dens_mass_particle - (1 - fs.solid_properties.particle_porosity)*fs.MB.solid_phase.properties[0.0,0.9].dens_mass_skeletal :   0.0 :   True


But this looks like a very reasonable constraint. After some thought, which admittedly requires some knowledge of the process we are modeling, we decide that the right approach is to make particle porosity a variable. We have assumed that porosity is constant, but this overconstrained subsystem is telling us that this assumption is not valid.

### How did we figure this out? (OKAY TO SKIP)
Adding a variable (including by unfixing a parameter) to an over-constraining constraint will often remove that constraint from the over-constrained subsystem. But how did we know that this was the right thing to do? If you just care using the diagnostics toolbox to extract as much information about a singularity as possible, you can skip this section. But if you are curious how we determined that particle porosity should not be constant, read on.

`dens_mass_skeletal` is determined purely by the composition of solid, which is made up of Fe<sub>2</sub>O<sub>3</sub>, Fe<sub>3</sub>O<sub>4</sub>, and inert Ti<sub>2</sub>O<sub>3</sub>. We can view the `density_skeletal_constraint` as follows:

In [11]:
model.fs.MB.solid_phase.properties[0, 0.9].density_skeletal_constraint.pprint()

density_skeletal_constraint : Size=1, Index=None, Active=True
    Key  : Lower : Body                                                                                                                                                                                                                                                                                                    : Upper : Active
    None :   1.0 : fs.MB.solid_phase.properties[0.0,0.9].dens_mass_skeletal*(0.00019047619047619048*fs.MB.solid_phase.properties[0.0,0.9].mass_frac_comp[Fe2O3] + 0.0002*fs.MB.solid_phase.properties[0.0,0.9].mass_frac_comp[Fe3O4] + 0.00025081514923501377*fs.MB.solid_phase.properties[0.0,0.9].mass_frac_comp[Al2O3]) :   1.0 :   True


If we assume a constant particle porosity, this gives us a particle porosity that is also uniquely determined by the solid composition by the above `density_particle_constraint`:
```
dens_mass_particle = (1 - porosity) * dens_mass_skeletal
```
But the composition of the solid is determined by the (somewhat misnamed) `material_holdup_calculation` constraints. While the name of these constraints implies they "calculate holdups," material holdups at $t=0$ are fixed as initial conditions (because holdups are the differential variables with respect to time in this model). At other time points, we assume that holdups are specified by differential and discretization equations of the model. This means that the `material_holdup_calculation` constraints actually calculate the solid phase mass fractions *from* the holdups. But as we hinted at above, the 4-by-4 system of holdup calculation constraints, `sum_component_eqn` (which simply constrains the sum of mass fractions to be one), mass fractions, and `dens_mass_particle`, uniquely solve for `dens_mass_particle` *as well as* the mass fractions. But if the holdup variables can be used to solve for the mass fractions, they *also* solve  for `dens_mass_skeletal`. So both sides of `density_particle_constraint` are already uniquely determined! This implies that we don't need this constraint at all, but we also know that this constraint has to hold. Something has to give. With this in mind, we actually have several options for how to resolve this overspecification:
1. Remove `density_particle_constraint`. Then we would have `dens_mass_particle` and `dens_mass_skeletal`, with no relationship between them. This would leave us with a mathematically sound model, but with densities that contradict constant particle porosity that we have assumed (which is used elsewhere in the reaction rate calculation equations).
2. Remove the constraints that calculate skeletal density from composition.
3. Relax particle porosity from a parameter to a variable.

Options 2 and 3 are equally valid. We've chosen option 3, meaning we assume that the particle "evolves" with a density that is well determined from its constituent species, rather than changing density to accommodate whatever mass it accumulates via reaction without altering its volume. This exercise should remind us that all mathematical modeling is somewhat of an art. In the process of choosing the "least bad" model, it is fairly easy to over or under-specify something by making the wrong combination of assumptions, and the Dulmage-Mendelsohn decomposition is a great tool for detecting when this has happened.

## Debugging the under-constrained subsystem

The under-constrained system does not decompose into independent subsystems, making it more difficult to debug. However, by inspection, we notice that the same constraints and variables seem to be repeated at each point in the length domain. For each point in space, the "odd variable out" seems to be the total flow rate `flow_mass`. Using some intuition about this particular process model, we may conclude that this variable should be calculated from the solid phase velocity, which is constant. We expect an equation that looks like
```
flow_mass == velocity * area * density
```

But this equation isn't here... so we need to add it.

## Fixing the model

We'll start by creating a fresh copy of the model, so we don't accidentally rely on IPOPT's point of termination.

In [12]:
model2 = make_model()
# Make the model square while we try to fix the structural singularity
model2.fs.MB.gas_phase.properties[:, 0].flow_mol.fix()
model2.fs.MB.solid_phase.properties[:, 1].flow_mass.fix()
model2.piecewise_constant_constraints.deactivate()

Constructing a steady model to initialize the dynamic model
Initializing steady model
2024-05-24 17:39:39 [INFO] idaes.init.fs.MB: Initialize Thermophysical Properties


2024-05-24 17:39:39 [INFO] idaes.init.fs.MB: Initialize Hydrodynamics


2024-05-24 17:39:39 [INFO] idaes.init.fs.MB: Initialize Mass Balances


2024-05-24 17:39:39 [INFO] idaes.init.fs.MB: Initialize Energy Balances


Ipopt 3.13.2: 

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Process Systems Engineering Framework (IDAES PSE
    Framework) Copyright (c) 2018-2019. See https://github.com/IDAES/idaes-pse.

This version of Ipopt was compiled using HSL, a collection of Fortran codes
    for large-scale scientific computation.  All technical papers, sales and
    publicity material resulting from use of the HSL codes within IPOPT must
    contain the following acknowledgement:
        HSL, a collection of Fortran codes for large-scale scientific
        computation. See http://

Solving square problem with dynamic model


Ipopt 3.13.2: 

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Process Systems Engineering Framework (IDAES PSE
    Framework) Copyright (c) 2018-2019. See https://github.com/IDAES/idaes-pse.

This version of Ipopt was compiled using HSL, a collection of Fortran codes
    for large-scale scientific computation.  All technical papers, sales and
    publicity material resulting from use of the HSL codes within IPOPT must
    contain the following acknowledgement:
        HSL, a collection of Fortran codes for large-scale scientific
        computation. See http://

Cannot recompute multipliers for feasibility problem.  Error in eq_mult_calculator

Number of Iterations....: 0

                                   (scaled)                 (unscaled)
Objective...............:   0.0000000000000000e+00    0.0000000000000000e+00
Dual infeasibility......:   0.0000000000000000e+00    0.0000000000000000e+00
Constraint violation....:   9.3496055342257023e-10    1.4901161193847656e-08
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   9.3496055342257023e-10    1.4901161193847656e-08


Number of objective function evaluations             = 1
Number of objective gradient evaluations             = 1
Number of equality constraint evaluations            = 1
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 1
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 0
Total CPU secs in IPOPT (w/o f

Initializing steady model
2024-05-24 17:39:41 [INFO] idaes.init.fs.MB: Initialize Thermophysical Properties


2024-05-24 17:39:41 [INFO] idaes.init.fs.MB: Initialize Hydrodynamics


2024-05-24 17:39:41 [INFO] idaes.init.fs.MB: Initialize Mass Balances


2024-05-24 17:39:41 [INFO] idaes.init.fs.MB: Initialize Energy Balances


Ipopt 3.13.2: 

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Process Systems Engineering Framework (IDAES PSE
    Framework) Copyright (c) 2018-2019. See https://github.com/IDAES/idaes-pse.

This version of Ipopt was compiled using HSL, a collection of Fortran codes
    for large-scale scientific computation.  All technical papers, sales and
    publicity material resulting from use of the HSL codes within IPOPT must
    contain the following acknowledgement:
        HSL, a collection of Fortran codes for large-scale scientific
        computation. See http://

Ipopt 3.13.2: 

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Process Systems Engineering Framework (IDAES PSE
    Framework) Copyright (c) 2018-2019. See https://github.com/IDAES/idaes-pse.

This version of Ipopt was compiled using HSL, a collection of Fortran codes
    for large-scale scientific computation.  All technical papers, sales and
    publicity material resulting from use of the HSL codes within IPOPT must
    contain the following acknowledgement:
        HSL, a collection of Fortran codes for large-scale scientific
        computation. See http://

## Adding a new particle porosity variable

In [13]:
model2.fs.MB.particle_porosity = pyo.Var(
    model2.fs.time, model2.fs.MB.length_domain, initialize=model2.fs.solid_properties.particle_porosity.value
)

Now we need to replace the old particle porosity parameter with this new variable. Luckily, the old parameter is actually implemented as a fixed variable, so we can easily identify all the constraints it participates in with `IncidenceGraphInterface`:

In [14]:
from pyomo.contrib.incidence_analysis import IncidenceGraphInterface

igraph = IncidenceGraphInterface(model2, include_fixed=True)
porosity_param = model2.fs.solid_properties.particle_porosity
print(f"Constraints containing {porosity_param.name}:")
for con in igraph.get_adjacent_to(porosity_param):
    print(f"  {con.name}")

Constraints containing fs.solid_properties.particle_porosity:
  fs.MB.solid_phase.properties[0.0,0.0].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,0.1].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,0.2].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,0.3].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,0.4].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,0.5].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,0.6].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,0.7].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,0.8].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,0.9].density_particle_constraint
  fs.MB.solid_phase.properties[0.0,1.0].density_particle_constraint
  fs.MB.solid_phase.properties[30.0,0.0].density_particle_constraint
  fs.MB.solid_phase.properties[30.0,0.1].density_particle_constraint
  fs.MB.solid_phase.properties[30.0,0.2].density_par

Particle porosity only appears in two constraints: the density constraint we saw above, and the reaction rate equation. We can replace particle porosity in these constraints using Pyomo's `replace_expressions` function:

In [15]:
from pyomo.core.expr import replace_expressions

for t, x in model2.fs.time * model2.fs.MB.length_domain:
    substitution_map = {id(porosity_param): model2.fs.MB.particle_porosity[t, x]}
    sp = model2.fs.MB.solid_phase
    cons = [sp.properties[t, x].density_particle_constraint, sp.reactions[t, x].gen_rate_expression["R1"]]
    for con in cons:
        con.set_value(replace_expressions(con.expr, substitution_map, descend_into_named_expressions=True))

We have added a new `particle_porosity` variable, and are using it in the relevant locations. Now we can move on to adding the missing constraint.

## Adding a new density-flow rate constraint

In [16]:
@model2.fs.MB.Constraint(model2.fs.time, model2.fs.MB.length_domain)
def density_flowrate_constraint(mb, t, x):
    return (
        mb.velocity_superficial_solid[t] * mb.bed_area
        * mb.solid_phase.properties[t, x].dens_mass_particle
        == mb.solid_phase.properties[t, x].flow_mass
    )

## Testing the new model

Let's see if these changes have fixed our model.

In [17]:
# Construct a new diagnostics toolbox
dt = DiagnosticsToolbox(model2)
dt.report_structural_issues()

Model Statistics

        Activated Blocks: 387 (Deactivated: 0)
        Free Variables in Activated Constraints: 10321 (External: 0)
            Free Variables with only lower bounds: 0
            Free Variables with only upper bounds: 0
            Free Variables with upper and lower bounds: 0
        Fixed Variables in Activated Constraints: 211 (External: 0)
        Activated Equality Constraints: 10321 (Deactivated: 10)
        Activated Inequality Constraints: 0 (Deactivated: 0)
        Activated Objectives: 1 (Deactivated: 0)

------------------------------------------------------------------------------------
2 WARNINGS


------------------------------------------------------------------------------------
1 Cautions

    Caution: 144 unused variables (4 fixed)

------------------------------------------------------------------------------------
Suggested next steps:

    display_components_with_inconsistent_units()
    display_potential_evaluation_errors()



The structural singularity seems to be gone! Let's unfix our degrees of freedom and see if we can solve.

In [18]:
model2.fs.MB.gas_phase.properties[:, 0].flow_mol.unfix()
model2.fs.MB.gas_phase.properties[0, 0].flow_mol.fix()
model2.fs.MB.solid_phase.properties[:, 1].flow_mass.unfix()
model2.fs.MB.solid_phase.properties[0, 1].flow_mass.fix()
model2.piecewise_constant_constraints.activate()

res = solver.solve(model2, tee=True)
print(f"Converged successfully: {pyo.check_optimal_termination(res)}")

Ipopt 3.13.2: max_iter=20
print_user_options=yes
option_file_name=/var/folders/j3/k8s7mg5x053c0rgp8sd87d540016lk/T/tmp5dzmb8mz_ipopt.opt



Using option file "/var/folders/j3/k8s7mg5x053c0rgp8sd87d540016lk/T/tmp5dzmb8mz_ipopt.opt".


List of user-set options:

                                    Name   Value                used
                                max_iter = 20                    yes
                        option_file_name = /var/folders/j3/k8s7mg5x053c0rgp8sd87d540016lk/T/tmp5dzmb8mz_ipopt.opt  yes
                       print_info_string = yes                   yes
                      print_user_options = yes                   yes

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Pro

   1  1.7521239e+04 8.45e+01 1.80e+03  -1.0 4.99e+02  -4.0 1.00e+00 1.25e-01f  4 Nh LTmaxTmaxTmaxTmax


   2  1.7535291e+04 5.53e+01 1.28e+05  -1.0 4.80e+00  -0.9 1.00e+00 1.00e+00h  1 LDj L


   3  1.7533314e+04 5.91e+01 9.39e+04  -1.0 2.85e+00   0.5 1.00e+00 1.00e+00h  1 l


   4  1.7531376e+04 6.18e+01 1.63e+05  -1.0 1.97e+00   0.9 1.00e+00 1.00e+00h  1 l


   5  1.7532039e+04 5.98e+01 2.44e+05  -1.0 7.29e+00   1.3 1.00e+00 2.50e-01h  3 lTmax


   6  1.7531943e+04 5.92e+01 1.07e+04  -1.0 8.42e-01   1.7 1.00e+00 1.00e+00h  1 l


   7  1.7531610e+04 5.78e+01 4.39e+03  -1.0 2.44e-01   1.3 1.00e+00 1.00e+00h  1 l


   8  1.7530730e+04 5.82e+01 1.35e+04  -1.0 3.83e-01   0.8 1.00e+00 1.00e+00f  1 l


   9  1.7530159e+04 7.00e+01 2.32e+05  -1.0 7.76e-01   0.3 1.00e+00 1.00e+00h  1 l


iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  10  1.7530383e+04 7.07e+01 2.15e+04  -1.0 2.31e-01   0.7 1.00e+00 1.00e+00h  1 l


  11  1.7532473e+04 7.05e+01 8.42e+03  -1.0 5.56e-01   0.3 1.00e+00 1.00e+00h  1 l


  12  1.7532758e+04 7.03e+01 8.16e+03  -1.0 1.85e+00  -0.2 1.00e+00 3.12e-02h  6 l


  13  1.7532965e+04 7.02e+01 8.09e+03  -1.0 5.55e+00  -0.7 1.00e+00 7.81e-03h  8 lTmax


  14  1.7533286e+04 7.00e+01 7.84e+03  -1.0 2.09e+00  -0.3 1.00e+00 3.12e-02h  6 l


  15  1.7533543e+04 6.97e+01 7.38e+03  -1.0 7.88e-01   0.2 1.00e+00 6.25e-02h  5 l


  16  1.7533908e+04 6.95e+01 7.14e+03  -1.0 2.37e+00  -0.3 1.00e+00 3.12e-02h  6 l


  17  1.7534200e+04 6.93e+01 6.71e+03  -1.0 8.93e-01   0.1 1.00e+00 6.25e-02h  5 l


  18  1.7534615e+04 6.91e+01 6.49e+03  -1.0 2.69e+00  -0.4 1.00e+00 3.12e-02h  6 l


  19  1.7534947e+04 6.88e+01 6.10e+03  -1.0 1.01e+00   0.1 1.00e+00 6.25e-02h  5 l


iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  20  1.7535213e+04 6.85e+01 5.41e+03  -1.0 3.84e-01   0.5 1.00e+00 1.25e-01h  4 l

Number of Iterations....: 20

                                   (scaled)                 (unscaled)
Objective...............:   1.7535212527023083e+04    1.7535212527023083e+04
Dual infeasibility......:   5.4114139743376509e+03    5.4114139743376509e+03
Constraint violation....:   7.6798115031223458e-03    6.8492626188812210e+01
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   3.3869144710429254e+01    5.4114139743376509e+03


Number of objective function evaluations             = 73
Number of objective gradient evaluations             = 21
Number of equality constraint evaluations            = 79
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 21
Number of inequality constraint Jacobian evaluations = 0
N

Converged successfully: False


This doesn't look much better. What's going on? I thought we just fixed the issue?

# Debugging the model, take two

Let's check the diagnostics toolbox for numerical issues?

In [19]:
model2.fs.MB.gas_phase.properties[:, 0].flow_mol.fix()
model2.fs.MB.solid_phase.properties[:, 1].flow_mass.fix()
model2.piecewise_constant_constraints.deactivate()
dt.report_numerical_issues()

2024-05-24 17:40:35 [INFO] idaes.core.util.model_diagnostics: Factor is exactly singular


Model Statistics

    Jacobian Condition Number: Undefined (Exactly Singular)

------------------------------------------------------------------------------------
4 WARNINGS


------------------------------------------------------------------------------------
6 Cautions

    Caution: 219 Variables with value close to zero (tol=1.0E-08)
    Caution: 1314 Variables with extreme value (<1.0E-04 or >1.0E+04)
    Caution: 99 Variables with None value
    Caution: 1619 Variables with extreme Jacobian values (<1.0E-04 or >1.0E+04)
    Caution: 1870 Constraints with extreme Jacobian values (<1.0E-04 or >1.0E+04)
    Caution: 3554 extreme Jacobian Entries (<1.0E-04 or >1.0E+04)

------------------------------------------------------------------------------------
Suggested next steps:

    display_constraints_with_large_residuals()
    display_variables_with_extreme_jacobians()
    display_constraints_with_extreme_jacobians()
    display_near_parallel_constraints()



Looks like we have "parallel constraints", which are another form of singularity. Let's follow the toolbox's advice to see what they are.

In [20]:
dt.display_near_parallel_constraints()

The following pairs of constraints are nearly parallel:

    fs.MB.solid_super_vel[0.0], fs.MB.density_flowrate_constraint[0.0,1.0]
    fs.MB.solid_super_vel[30.0], fs.MB.density_flowrate_constraint[30.0,1.0]
    fs.MB.solid_super_vel[60.0], fs.MB.density_flowrate_constraint[60.0,1.0]
    fs.MB.solid_super_vel[90.0], fs.MB.density_flowrate_constraint[90.0,1.0]
    fs.MB.solid_super_vel[120.0], fs.MB.density_flowrate_constraint[120.0,1.0]
    fs.MB.solid_super_vel[150.0], fs.MB.density_flowrate_constraint[150.0,1.0]
    fs.MB.solid_super_vel[180.0], fs.MB.density_flowrate_constraint[180.0,1.0]
    fs.MB.solid_super_vel[210.0], fs.MB.density_flowrate_constraint[210.0,1.0]
    fs.MB.solid_super_vel[240.0], fs.MB.density_flowrate_constraint[240.0,1.0]
    fs.MB.solid_super_vel[270.0], fs.MB.density_flowrate_constraint[270.0,1.0]
    fs.MB.solid_super_vel[300.0], fs.MB.density_flowrate_constraint[300.0,1.0]



`density_flowrate_constraint` is the constraint that we added. What is `solid_super_vel`?

In [21]:
model2.fs.MB.solid_super_vel[0].pprint()

{Member of solid_super_vel} : Solid superficial velocity
    Size=11, Index=fs._time, Active=True
    Key : Lower : Body                                                                                                                                                            : Upper : Active
    0.0 :   0.0 : fs.MB.velocity_superficial_solid[0.0]*fs.MB.bed_area*fs.MB.solid_phase.properties[0.0,1.0].dens_mass_particle - fs.MB.solid_phase.properties[0.0,1.0].flow_mass :   0.0 :   True


This is the same as the constraint we just added! Looks like that constraint already existed at the solid inlet. We can easily deactivate the new constraints at this point in the length domain:

In [22]:
model2.fs.MB.density_flowrate_constraint[:, 1.0].deactivate();

But now we have removed constraints from a square model, and expect to have degrees of freedom. Let's see what the diagnostics toolbox has to say.

In [23]:
dt = DiagnosticsToolbox(model2)
dt.report_structural_issues()

Model Statistics

        Activated Blocks: 387 (Deactivated: 0)
        Free Variables in Activated Constraints: 10321 (External: 0)
            Free Variables with only lower bounds: 0
            Free Variables with only upper bounds: 0
            Free Variables with upper and lower bounds: 0
        Fixed Variables in Activated Constraints: 211 (External: 0)
        Activated Equality Constraints: 10310 (Deactivated: 21)
        Activated Inequality Constraints: 0 (Deactivated: 0)
        Activated Objectives: 1 (Deactivated: 0)

------------------------------------------------------------------------------------
4 WARNINGS

        Under-Constrained Set: 8881 variables, 8870 constraints
        Over-Constrained Set: 0 variables, 0 constraints

------------------------------------------------------------------------------------
1 Cautions

    Caution: 144 unused variables (4 fixed)

------------------------------------------------------------------------------------
Suggested nex

But this doesn't help us very much. We have some extraneous degrees of freedom, but with 8881 variables in the under-constrained subsystem, it will be difficult to tell what they are. After some thought (and model-specific intuition), we land on the conclusion that maybe we need to fix particle porosity at the solid inlet. Here, total flow rate is specified, and the `solid_super_vel` equation is using it to compute velocity. So we need `dens_mass_particle` to be known, which means we need `particle_porosity` to be fixed.

In [24]:
model2.fs.MB.particle_porosity[:, 1.0].fix();

Let's run the diagnostics toolbox as a sanity check.

In [25]:
dt = DiagnosticsToolbox(model2)
dt.report_structural_issues()
dt.report_numerical_issues()

Model Statistics

        Activated Blocks: 387 (Deactivated: 0)
        Free Variables in Activated Constraints: 10310 (External: 0)
            Free Variables with only lower bounds: 0
            Free Variables with only upper bounds: 0
            Free Variables with upper and lower bounds: 0
        Fixed Variables in Activated Constraints: 222 (External: 0)
        Activated Equality Constraints: 10310 (Deactivated: 21)
        Activated Inequality Constraints: 0 (Deactivated: 0)
        Activated Objectives: 1 (Deactivated: 0)

------------------------------------------------------------------------------------
2 WARNINGS


------------------------------------------------------------------------------------
1 Cautions

    Caution: 144 unused variables (4 fixed)

------------------------------------------------------------------------------------
Suggested next steps:

    display_components_with_inconsistent_units()
    display_potential_evaluation_errors()



Model Statistics

    Jacobian Condition Number: 7.617E+25

------------------------------------------------------------------------------------
3 WARNINGS


------------------------------------------------------------------------------------
6 Cautions

    Caution: 219 Variables with value close to zero (tol=1.0E-08)
    Caution: 1314 Variables with extreme value (<1.0E-04 or >1.0E+04)
    Caution: 99 Variables with None value
    Caution: 1619 Variables with extreme Jacobian values (<1.0E-04 or >1.0E+04)
    Caution: 1859 Constraints with extreme Jacobian values (<1.0E-04 or >1.0E+04)
    Caution: 3543 extreme Jacobian Entries (<1.0E-04 or >1.0E+04)

------------------------------------------------------------------------------------
Suggested next steps:

    display_constraints_with_large_residuals()
    display_variables_with_extreme_jacobians()
    display_constraints_with_extreme_jacobians()



Looks good! Now we can release our degrees of freedom and try to solve again.

In [26]:
model2.fs.MB.gas_phase.properties[:, 0].flow_mol.unfix()
model2.fs.MB.gas_phase.properties[0, 0].flow_mol.fix()
model2.fs.MB.solid_phase.properties[:, 1].flow_mass.unfix()
model2.fs.MB.solid_phase.properties[0, 1].flow_mass.fix()
model2.piecewise_constant_constraints.activate()

res = solver.solve(model2, tee=True)
print(f"Converged successfully: {pyo.check_optimal_termination(res)}")

Ipopt 3.13.2: max_iter=20
print_user_options=yes
option_file_name=/var/folders/j3/k8s7mg5x053c0rgp8sd87d540016lk/T/tmp3_4po4n__ipopt.opt



Using option file "/var/folders/j3/k8s7mg5x053c0rgp8sd87d540016lk/T/tmp3_4po4n__ipopt.opt".


List of user-set options:

                                    Name   Value                used
                                max_iter = 20                    yes
                        option_file_name = /var/folders/j3/k8s7mg5x053c0rgp8sd87d540016lk/T/tmp3_4po4n__ipopt.opt  yes
                       print_info_string = yes                   yes
                      print_user_options = yes                   yes

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt

This version of Ipopt was compiled from source code available at
    https://github.com/IDAES/Ipopt as part of the Institute for the Design of
    Advanced Energy Systems Pro

   1  4.4256949e+03 1.01e+03 1.33e+07  -1.0 3.98e+05    -  1.00e+00 1.00e+00f  1 


   2  4.4845871e+03 1.00e+00 5.34e+04  -1.0 6.13e+03    -  1.00e+00 1.00e+00h  1 Nhj 


   3  4.4844740e+03 2.80e-04 2.63e+01  -1.0 2.07e+02    -  1.00e+00 1.00e+00h  1 


   4  4.4844740e+03 2.98e-08 2.51e-07  -5.7 5.10e-03    -  1.00e+00 1.00e+00h  1 

Number of Iterations....: 4

                                   (scaled)                 (unscaled)
Objective...............:   4.4844740157744627e+03    4.4844740157744627e+03
Dual infeasibility......:   2.5136250769719481e-07    2.5136250769719481e-07
Constraint violation....:   1.5279510989785194e-09    2.9802322387695312e-08
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   1.5279510989785194e-09    2.5136250769719481e-07


Number of objective function evaluations             = 5
Number of objective gradient evaluations             = 5
Number of equality constraint evaluations            = 5
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 5
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 4
Total CPU secs in IPOPT (w/o fu

Converged successfully: True


It worked! For the simple optimization problem we have set up, this solve looks a lot more like what we expect.

## Takeaways from this tutorial
What have we learned?
1. IPOPT using non-zero regularization coefficients hints at a singular Jacobian (especially when "L"/"l" diagnostic tags are present).
2. When this happens, start by calling `report_structural_issues` to check for a structural singularity. If this looks good, call `report_numerical_issues` to check for a numerical singularity.
3. When debugging a structural singularity, decomposing a problem into subsystems that each should be nonsingular (e.g. unit models or points in time) is very useful.
4. The solution to a structural singularity is often to relax a fixed parameter, add a constraint that was forgotten, remove a constraint that was redundant, or fix an extraneous degree of freedom.
5. Model-specific intuition is usually necessary to diagnose and fix modeling issues. (If you're an algorithm developer, learn about the models you're using! If you don't understand your models, you don't understand your algorithms!)
6. A modeling issue doesn't necessarily have a unique solution. This is especially true when the issue involves invalid assumptions.
7. Debugging is an iterative process &mdash; fixing one issue can introduce another.

### References

[[1]] Okoli et al., "A framework for the optimization of chemical looping combustion processes". *Powder Tech*, 2020.

[[2]] Parker and Biegler, "Dynamic modeling and nonlinear model predictive control of a moving bed chemical looping combustion reactor". *IFAC PapersOnline*, 2022.

[[3]] Dulmage and Mendelsohn, "Coverings of bipartite graphs". *Can J. Math.*, 1958.

[[4]] Pothen and Fan, "Computing the block triangular form of a sparse matrix". *ACM Trans. Math. Softw.*, 1990.

[[5]] Parker et al., "Applications of the Dulmage-Mendelsohn decomposition for debugging nonlinear optimization problems". *Comp. Chem. Eng.*, 2023.

[1]: https://www.sciencedirect.com/science/article/pii/S0032591019302803
[2]: https://www.sciencedirect.com/science/article/pii/S2405896322008825
[3]: https://www.cambridge.org/core/journals/canadian-journal-of-mathematics/article/coverings-of-bipartite-graphs/413735C5888AB542B92D0C4F402800B1
[4]: https://dl.acm.org/doi/10.1145/98267.98287
[5]: https://www.sciencedirect.com/science/article/pii/S0098135423002533
